src/util/records.py

```diff
@@ -1504,6 +1512,30 @@ def replace_record_references_batch(cr, id_mapping, model_src, model_dst=None, r
                     [list(map(list, id_mapping.items()))],
                 ).decode()
                 explode_execute(cr, query, table=ir.table)
+                # ensure all new ids exist
+                cr.execute(
+                    format_query(
+                        cr,
+                        """
+                        Select t.id,
+                               j.key,
+                               j.value
+                          FROM {table} t
+                          JOIN JSONB_EACH(t.{column}) j
+                            ON True
+                         WHERE NOT EXISTS (
+                                    SELECT 1 FROM {dest_table} WHERE id = j.value::int
+                               )
+                        """,
+                        table=ir.table,
+                        column=ir.res_id,
+                        dest_table=table_of_model(cr, model_dst),
+                    )
+                )
+                invalid_ref = list(cr.fetchall())
+                assert not invalid_ref, "Invalid company dependent values for {}.{} referencing model {}: {}".format(
+                    model_of_table(cr, ir.table), ir.res_id, model_dst, invalid_ref
+                )
```

Comment:
```suggestion
                if invalid_ref:
                    raise RuntimeError("Invalid company dependent values for {}.{} referencing model {}: {}".format(model_of_table(cr, ir.table), ir.res_id, model_dst, invalid_ref))
```

NOTE: the suggestion will still need to be formatted correctly.